### PR TITLE
Martin/ios 458 conversation list gets stuck between

### PR DESCRIPTION
**Pull request by MartinJibber**
Fixing issues where the conversations weren't centered.
Fixes to when the accessory view/keyboard shows or hides.
Also added a delay before auto-showing/hiding the accessory view.

https://user-images.githubusercontent.com/88675954/154379311-24f360bc-7de5-43f9-9b8c-2437777fa58d.MP4

### CODE REVIEW

File: Shared/Managers/KeyboardManager.swift [hunk 1]
```diff
@@ -9,25 +9,6 @@
...
-                                        for: nil)
-        return UIResponder.currentFirstResponder
-    }
-
-    @objc private func findFirstResponder(sender: AnyObject) {
-        UIResponder.currentFirstResponder = self
-    }
-}
-
-
```
Comment (MartinJibber) [inline Shared/Managers/KeyboardManager.swift:30]: Moved this out into its own extension.

File: Shared/Managers/KeyboardManager.swift [hunk 2]
```diff
@@ -124,7 +105,7 @@ class KeyboardManager {
 
     private func getInputAccessoryHeight() -> CGFloat {
         var inputAccessoryHeight: CGFloat = 0
-        if let responder = UIResponder.firstReponder,
+        if let responder = UIResponder.firstResponder,
```
Comment (MartinJibber) [inline Shared/Managers/KeyboardManager.swift:108]: Whoops, "Responder" not "Reponder"

File: iOS/Extensions/UIResponder+Extensions.swift [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    private weak static var currentFirstResponder: UIResponder?
+
+    /// Returns the UIResponder that is currently designated as first responder. Nil is returned if there is no current first responder.
+    static var firstResponder: UIResponder? {
+        UIResponder.currentFirstResponder = nil
+        UIApplication.shared.sendAction(#selector(findFirstResponder(sender:)),
+                                        to: nil,
+                                        from: nil,
+                                        for: nil)
+        return UIResponder.currentFirstResponder
```
Comment (MartinJibber) [inline iOS/Extensions/UIResponder+Extensions.swift:22]: See comment

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 4]
```diff
@@ -51,7 +51,7 @@ class ConversationListViewController: ViewController, ConversationListCollection
     }
 
     override var canBecomeFirstResponder: Bool {
-        return self.presentedViewController.isNil
+        return self.presentedViewController.isNil && ConversationsManager.shared.activeConversation.exists
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:54]: We never want to show the accessory view if there is no active conversation. 

Correct me if I'm wrong.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 5]
```diff
@@ -303,14 +301,31 @@ class ConversationListViewController: ViewController, ConversationListCollection
                 self?.messageInputAccessoryView.textView.setPlaceholder(for: users, isReply: false)
                 self?.messageInputAccessoryView.updateSwipeHint(shouldPlay: true)
             }
-        } else {
-            ConversationsManager.shared.activeConversation = nil
 
-            if self.isFirstResponder {
-                self.resignFirstResponder()
+            self.firstResponderTask = Task { [weak self] in
+                await Task.sleep(seconds: 0.25)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:306]: Adding a delay before we become first responder so it doesn't feel so abrupt.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 6]
```diff
@@ -303,14 +301,31 @@ class ConversationListViewController: ViewController, ConversationListCollection
...
+
+                guard !Task.isCancelled else { return }
+
+                // The input accessory view should be shown when centered on a conversation. If there's not
+                // already set as first responder, then make the VC first responder.
+                if UIResponder.firstResponder.isNil {
+                    self?.becomeFirstResponder()
+                }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:314]: When the keyboard is up, the text field is actually the first responder, which is why we can't just check ourselves.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 7]
```diff
@@ -303,14 +301,31 @@ class ConversationListViewController: ViewController, ConversationListCollection
...
             }
+        } else {
+            ConversationsManager.shared.activeConversation = nil
 
             self.messageInputAccessoryView.updateSwipeHint(shouldPlay: true)
+
+            self.firstResponderTask = Task {
+                await Task.sleep(seconds: 0.25)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:322]: Same delay for hiding the keyboard.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 8]
```diff
@@ -303,14 +301,31 @@ class ConversationListViewController: ViewController, ConversationListCollection
...
+
+                guard !Task.isCancelled else { return }
+
+                // Hide the keyboard and accessory view when we're not centered on a conversation.
+                UIResponder.firstResponder?.resignFirstResponder()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:327]: The textfield might be the first responder so I have to resign whatever is current.

File: iOS/Flows/Conversation/ConversationListCollectionViewLayout.swift [hunk 9]
```diff
@@ -123,13 +123,6 @@ class ConversationListCollectionViewLayout: UICollectionViewFlowLayout {
         return CGPoint(x: newXOffset, y: proposedContentOffset.y)
     }
 
-    override func targetContentOffset(forProposedContentOffset proposedContentOffset: CGPoint) -> CGPoint {
-        self.sendCenterUpdateEventIfNeeded(withContentOffset: proposedContentOffset)
-
-        return self.targetContentOffset(forProposedContentOffset: proposedContentOffset,
-                                        withScrollingVelocity: .zero)
-    }
-
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationListCollectionViewLayout.swift:132]: Removing this fixed a lot of the issues with the conversation not being centered. 
It seems that overriding this causes weird issues when the collection view both scrolling and animating a layout change.

File: Shared/Managers/KeyboardManager.swift (hunk 2 above)
Comment (Benjamindodgson) [reply Shared/Managers/KeyboardManager.swift:108]: You don't want to ponder on this responder?

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**